File: ext/internal/features/amqp/consume_state.go
package amqp_feature

import (
	"context"
	"sconcur/internal/contracts"
	"sconcur/internal/dto"
	"sconcur/internal/features/amqp/payloads"
	"sconcur/internal/helpers"
	"sconcur/internal/states"
	"sconcur/internal/tasks"
	"strconv"
	"sync"
	"sync/atomic"
	"time"

	amqp091 "github.com/rabbitmq/amqp091-go"
	"github.com/vmihailenco/msgpack/v5"
)

var _ contracts.StateContract = (*consumeState)(nil)

// consumerCounter backs the generated consumer tags. The tag is generated here rather
// than left to the driver because the driver keeps the one it generates to itself, and
// PHP needs it to cancel the consumer and to route deliveries.
var consumerCounter atomic.Int64

// consumeState streams one consumer to PHP: the first Next returns the consumer tag, and
// every following one returns the next delivery. The stream ends when the consumer is
// cancelled, the channel or connection dies, the flow stops, or the read timeout expires.
type consumeState struct {
	mutex       sync.Mutex
	ctx         context.Context
	message     *dto.Message
	entry       *channelEntry
	consumerTag string
	readTimeout time.Duration
	startTime   time.Time
	metaSent    bool
	autoAck     bool
	deliveries  <-chan amqp091.Delivery
	cleanup     func()
}

func (s *consumeState) Next() *dto.Result {
	s.mutex.Lock()

	if !s.metaSent {
		s.metaSent = true

		s.mutex.Unlock()

		serialized, err := msgpack.Marshal(payloads.ConsumerMeta{ConsumerTag: s.consumerTag})

		if err != nil {
			return dto.NewErrorResult(s.message, errFactory.ByErr("marshal consumer meta", err))
		}

		return dto.NewSuccessResultWithNext(s.message, string(serialized), helpers.CalcExecutionMs(s.startTime))
	}

	s.mutex.Unlock()

	// Prefer a delivery that is already buffered over shutdown. With both a delivery and
	// a cancelled context ready, a plain select could pick the cancellation at random and
	// drop a message that had already arrived.
	select {
	case delivery, ok := <-s.deliveries:
		return s.resultFromDelivery(delivery, ok)
	default:
	}

	var deadline <-chan time.Time

	if s.readTimeout > 0 {
		timer := time.NewTimer(s.readTimeout)
		defer timer.Stop()

		deadline = timer.C
	}

	select {
	case delivery, ok := <-s.deliveries:
		return s.resultFromDelivery(delivery, ok)
	case <-deadline:
		// ext-amqp ends the consume loop with this exact failure when read_timeout
		// passes with no delivery, wording included.
		return dto.NewErrorResult(s.message, errorPayload(scopeCommand, 0, "Consumer timeout exceed"))
	case <-s.ctx.Done():
		return dto.NewSuccessResult(s.message, "", helpers.CalcExecutionMs(s.startTime))
	}
}

// resultFromDelivery turns a receive into a Next() result: one delivery with more to come,
// or the end of the stream once the driver closed the deliveries channel.
//
// The driver closes it both when the flow is going away and when the consumer was taken
// from us — the queue was deleted, the channel died, the broker cancelled the consumer.
// The first is the normal end of a consume loop; the second is a failure, and ext-amqp
// raises there, so the stream ends with an error rather than a quiet return.
func (s *consumeState) resultFromDelivery(delivery amqp091.Delivery, ok bool) *dto.Result {
	if !ok {
		if s.ctx.Err() != nil {
			return dto.NewSuccessResult(s.message, "", helpers.CalcExecutionMs(s.startTime))
		}

		// A connection that died ends every consumer on it the same way the broker
		// cancelling one does, and a worker told only that its consumer was cancelled
		// would open another on a connection that is not there. Scoped as the connection
		// failing, so PHP reports it as one.
		if s.entry.connectionClosed() {
			return dto.NewErrorResult(s.message, networkErrorPayload(
				"Consumer "+s.consumerTag+" ended: no connection available.",
			))
		}

		return dto.NewErrorResult(s.message, errorPayload(
			scopeCommand,
			0,
			"Consumer "+s.consumerTag+" was cancelled by the broker.",
		))
	}

	serialized, err := msgpack.Marshal(deliveryToPayload(delivery, s.entry.id))

	if err != nil {
		return dto.NewErrorResult(s.message, errFactory.ByErr("marshal delivery", err))
	}

	// Counted here, where the delivery leaves for PHP: the acknowledgement that
	// settles it arrives as an ordinary command, so the pair needs nothing extra on
	// the wire (see consumerstats.go).
	consumerStatsInstance.deliveryDispatched(s.entry.id, delivery.DeliveryTag, s.autoAck)

	return dto.NewSuccessResultWithNext(s.message, string(serialized), helpers.CalcExecutionMs(s.startTime))
}

// Close cancels the consumer, leaving the channel open — a channel outlives its
// consumers. It runs on a fresh context: by the time cleanup runs, the task context is
// already cancelled.
func (s *consumeState) Close() {
	s.cleanup()
}

// handleConsume registers a consumer and streams its deliveries. The consumer is not tied
// to the task context — the stream state owns it, and Close() cancels it when PHP stops
// reading or the flow ends.
func (f *AmqpFeature) handleConsume(task *tasks.Task, raw msgpack.RawMessage) {
	startTime := time.Now()
	message := task.GetMessage()

	entry, params, ok := resolveChannel[payloads.ConsumeParams](task, raw, "consume")

	if !ok {
		return
	}

	consumerTag := params.ConsumerTag

	if consumerTag == "" {
		consumerTag = nextConsumerTag()
	}

	registerContext, cancelRegister := commandContext(task, params.TimeoutMs, defaultRpcTimeout)
	defer cancelRegister()

	deliveries, err := entry.consume(registerContext, consumerTag, params)

	if err != nil {
		fail(task, entry, "consume", err)

		return
	}

	var cleanupOnce sync.Once

	cleanup := func() {
		cleanupOnce.Do(func() {
			consumerStatsInstance.consumerClosed(entry.id, consumerTag)

			// The basic.cancel goes through the channel's own lock, and only if this
			// consumer is still registered — PHP may have cancelled it already. That also
			// drops it from the registry, so the idle sweeper sees the channel as idle.
			entry.cancelConsumer(consumerTag)
		})
	}

	entry.registerConsumer(consumerTag, message.TaskKey)

	consumerStatsInstance.consumerOpened(entry.id, consumerTag)

	startConsumerTelemetry()

	state := &consumeState{
		ctx:         task.GetContext(),
		autoAck:     params.AutoAck,
		message:     message,
		entry:       entry,
		consumerTag: consumerTag,
		readTimeout: time.Duration(max(params.ReadTimeoutMs, 0)) * time.Millisecond,
		startTime:   startTime,
		deliveries:  deliveries,
		cleanup:     cleanup,
	}

	// The first Next returns the consumer tag with HasNext set, so the state stays alive
	// and the deliveries can be pulled through next(). states.Start hooks Close on flow
	// stop.
	result, err := states.Get().Start(task.GetContext(), message.TaskKey, state)

	if err != nil {
		cleanup()

		fail(task, entry, "consume", err)

		return
	}

	task.AddResult(result)
}

// consume registers the consumer on the driver channel, bounded by the command context
// like every other method. The consumer itself is not tied to a context: it is ended by
// cancelConsumer or by the channel closing, both of which are serialized against the other
// commands of this channel.
func (e *channelEntry) consume(
	ctx context.Context,
	consumerTag string,
	params payloads.ConsumeParams,
) (<-chan amqp091.Delivery, error) {
	arguments := mapToTable(params.Arguments)

	var deliveries <-chan amqp091.Delivery

	err := e.doAbandoning(
		ctx,
		func(channel *amqp091.Channel) error {
			var consumeError error

			deliveries, consumeError = channel.ConsumeWithContext(
				context.Background(),
				params.QueueName,
				consumerTag,
				params.AutoAck,
				params.Exclusive,
				params.NoLocal,
				params.NoWait,
				arguments,
			)

			return consumeError
		},
		// The registration outran its deadline: PHP was told it failed and will never
		// read this consumer, but the broker has one and will keep feeding it. Nothing
		// else can cancel it — entry.consumers never learned the tag — so it is
		// cancelled here, or the queue quietly stops making progress at its prefetch.
		func(consumeError error) {
			if consumeError == nil {
				e.sendCancel(consumerTag, false)
			}
		},
	)

	if err != nil {
		return nil, err
	}

	return deliveries, nil
}

func nextConsumerTag() string {
	return "sconcur-ctag-" + strconv.FormatInt(consumerCounter.Add(1), 10)
}

File: ext/src/features/amqp/consume_state.rs
//! One consumer streamed to
//! PHP through the shared streaming-state registry.
//!
//! The first `next()` returns the consumer tag, and every following one returns
//! the next delivery. The stream ends when the consumer is cancelled, the
//! channel or connection dies, the flow stops, or the read timeout expires.

use std::sync::Arc;
use std::sync::atomic::{AtomicBool, AtomicI64, Ordering};
use std::time::{Duration, Instant};

use futures_util::StreamExt;
use lapin::Consumer;
use lapin::options::BasicConsumeOptions;
use tokio_util::sync::CancellationToken;

use crate::dto::{Message, Result};
use crate::helpers::calc_execution_ms;
use crate::states::{self, StateCloseFuture, StateContract, StateFuture};
use crate::tasks::Task;

use super::channels::{ChannelEntry, CommandError};
use super::connections::{DEFAULT_RPC_TIMEOUT, ms_or_default};
use super::consumerstats;
use super::payloads::{self, ChannelCommand};
use super::{SCOPE_COMMAND, error_payload, fail, network_error_payload};

/// Backs the generated consumer tags. The tag is generated here rather than
/// left to the driver because PHP needs it to cancel the consumer and to route
/// deliveries, and lapin keeps a generated one to itself.
static CONSUMER_COUNTER: AtomicI64 = AtomicI64::new(0);

pub fn next_consumer_tag() -> String {
    format!(
        "sconcur-ctag-{}",
        CONSUMER_COUNTER.fetch_add(1, Ordering::SeqCst) + 1
    )
}

struct ConsumeState {
    message: Arc<Message>,
    entry: Arc<ChannelEntry>,
    consumer_tag: String,
    read_timeout: Duration,
    start_time: Instant,
    auto_ack: bool,
    flow: CancellationToken,
    meta_sent: AtomicBool,
    cleaned: AtomicBool,
    consumer: tokio::sync::Mutex<Consumer>,
}

impl StateContract for ConsumeState {
    fn next(&self) -> StateFuture<'_> {
        Box::pin(async move {
            if !self.meta_sent.swap(true, Ordering::SeqCst) {
                return Result::success_with_next(
                    &self.message,
                    payloads::encode_consumer_meta(&self.consumer_tag),
                    calc_execution_ms(self.start_time),
                );
            }

            let mut consumer = self.consumer.lock().await;

            // A delivery that is already buffered wins over shutdown: with both
            // ready, an unbiased select could pick the cancellation at random
            // and drop a message that had already arrived.
            let received = if self.read_timeout.is_zero() {
                tokio::select! {
                    biased;

                    delivery = consumer.next() => Some(delivery),
                    _ = self.flow.cancelled() => None,
                }
            } else {
                tokio::select! {
                    biased;

                    delivery = consumer.next() => Some(delivery),
                    _ = tokio::time::sleep(self.read_timeout) => {
                        // ext-amqp ends the consume loop with this exact failure
                        // when read_timeout passes with no delivery, wording
                        // included.
                        return Result::error(
                            &self.message,
                            error_payload(SCOPE_COMMAND, 0, "Consumer timeout exceed"),
                        );
                    }
                    _ = self.flow.cancelled() => None,
                }
            };

            match received {
                Some(Some(Ok(delivery))) => {
                    // Counted here, where the delivery leaves for PHP: the
                    // acknowledgement that settles it arrives as an ordinary
                    // command, so the pair needs nothing extra on the wire.
                    consumerstats::stats().delivery_dispatched(
                        &self.entry.id,
                        delivery.delivery_tag,
                        self.auto_ack,
                    );

                    Result::success_with_next(
                        &self.message,
                        payloads::encode_delivery(&payloads::DeliveryOut {
                            channel_id: &self.entry.id,
                            consumer_tag: &self.consumer_tag,
                            delivery_tag: delivery.delivery_tag,
                            redelivered: delivery.redelivered,
                            exchange_name: delivery.exchange.as_str(),
                            routing_key: delivery.routing_key.as_str(),
                            body: &delivery.data,
                            properties: &delivery.properties,
                        }),
                        calc_execution_ms(self.start_time),
                    )
                }
                Some(Some(Err(error))) => {
                    self.entry.note_protocol_failure(&error);

                    self.ended()
                }
                // The driver ended the stream: the consumer was cancelled, the
                // channel died, or the queue was deleted.
                Some(None) => self.ended(),
                // The flow is going away, which is the normal end of a consume
                // loop.
                None => Result::success(
                    &self.message,
                    Vec::new(),
                    calc_execution_ms(self.start_time),
                ),
            }
        })
    }

    /// Cancels the consumer, leaving the channel open — a channel outlives its
    /// consumers. It runs on a deadline of its own: by the time cleanup runs,
    /// the task's own is long gone.
    fn close(&self) -> StateCloseFuture<'_> {
        Box::pin(async move {
            if self.cleaned.swap(true, Ordering::SeqCst) {
                return;
            }

            consumerstats::stats().consumer_closed(&self.entry.id, &self.consumer_tag);

            // Only if this consumer is still registered — PHP may have
            // cancelled it already. That also drops it from the registry, so the
            // idle sweeper sees the channel as idle.
            self.entry.cancel_consumer(&self.consumer_tag).await;
        })
    }
}

impl ConsumeState {
    /// The stream ended without the flow going away, which is a failure:
    /// ext-amqp raises there, so the stream ends with an error rather than a
    /// quiet return.
    fn ended(&self) -> Result {
        if self.flow.is_cancelled() {
            return Result::success(
                &self.message,
                Vec::new(),
                calc_execution_ms(self.start_time),
            );
        }

        // A connection that died ends every consumer on it the same way the
        // broker cancelling one does, and a worker told only that its consumer
        // was cancelled would open another on a connection that is not there.
        if self.entry.connection_closed() {
            return Result::error(
                &self.message,
                network_error_payload(&format!(
                    "Consumer {} ended: no connection available.",
                    self.consumer_tag
                )),
            );
        }

        Result::error(
            &self.message,
            error_payload(
                SCOPE_COMMAND,
                0,
                &format!(
                    "Consumer {} was cancelled by the broker.",
                    self.consumer_tag
                ),
            ),
        )
    }
}

/// Registers a consumer and streams its deliveries. The consumer is not tied to
/// the task's deadline — the stream state owns it, and `close()` cancels it when
/// PHP stops reading or the flow ends.
pub async fn handle_consume(task: &Task, raw: &rmpv::Value) {
    let Some((entry, params, start_time)) =
        super::commands::resolve::<payloads::ConsumeParams>(task, raw, "consume").await
    else {
        return;
    };

    let message = task.message();

    let consumer_tag = if params.consumer_tag.is_empty() {
        next_consumer_tag()
    } else {
        params.consumer_tag.clone()
    };

    let timeout = ms_or_default(params.timeout_ms(), DEFAULT_RPC_TIMEOUT);

    let consumer = match open_consumer(
        &entry,
        &consumer_tag,
        &params.queue_name,
        params.auto_ack,
        params.exclusive,
        params.no_local,
        params.no_wait,
        super::commands::arguments(&params.arguments),
        timeout,
    )
    .await
    {
        Ok(consumer) => consumer,
        Err(error) => {
            fail(task, Some(&entry), "consume", error).await;

            return;
        }
    };

    entry.register_consumer(&consumer_tag, &message.task_key);

    consumerstats::stats().consumer_opened(&entry.id, &consumer_tag);
    consumerstats::start_telemetry();

    let state = Arc::new(ConsumeState {
        message: task.message_arc(),
        entry: entry.clone(),
        consumer_tag: consumer_tag.clone(),
        read_timeout: Duration::from_millis(params.read_timeout_ms.max(0) as u64),
        start_time,
        auto_ack: params.auto_ack,
        flow: task.context().clone(),
        meta_sent: AtomicBool::new(false),
        cleaned: AtomicBool::new(false),
        consumer: tokio::sync::Mutex::new(consumer),
    });

    // The first next() returns the consumer tag with the more-coming flag, so
    // the state stays alive and the deliveries can be pulled through next().
    // states::start hooks close() on flow stop.
    match states::get()
        .start(task.context().clone(), &message.task_key, state.clone())
        .await
    {
        Ok(result) => task.add_result(result).await,
        Err(error) => {
            state.close().await;

            fail(
                task,
                Some(&entry),
                "consume",
                CommandError::Message(error),
            )
            .await;
        }
    }
}

/// Registers the consumer on the driver channel, bounded by the command
/// deadline like every other method.
///
/// A registration that outran its deadline is the one place a dropped future is
/// not enough: the broker may have taken the `basic.consume` anyway and will
/// keep feeding it, and nothing else can cancel it — the channel's registry
/// never learned the tag. So the cancel is sent here, or the queue quietly
/// stops making progress at its prefetch.
#[allow(clippy::too_many_arguments)]
pub async fn open_consumer(
    entry: &Arc<ChannelEntry>,
    consumer_tag: &str,
    queue_name: &str,
    auto_ack: bool,
    exclusive: bool,
    no_local: bool,
    no_wait: bool,
    arguments: lapin::types::FieldTable,
    timeout: Duration,
) -> std::result::Result<Consumer, CommandError> {
    let outcome = entry
        .run(
            timeout,
            entry.channel().basic_consume(
                super::commands::short(queue_name),
                super::commands::short(consumer_tag),
                BasicConsumeOptions {
                    no_local,
                    no_ack: auto_ack,
                    exclusive,
                    nowait: no_wait,
                },
                arguments,
            ),
        )
        .await;

    if matches!(outcome, Err(CommandError::CommandTimeout)) {
        let entry = entry.clone();
        let consumer_tag = consumer_tag.to_string();

        crate::core::get().runtime().spawn(async move {
            entry.send_cancel(&consumer_tag, false).await;
        });
    }

    outcome
}
